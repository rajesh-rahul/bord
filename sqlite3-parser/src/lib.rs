//! This is a parser to parse SQLite3(v.3.46.0 onwards). It follows the excellent error
//! resillient parsing [guide]((https://matklad.github.io/2023/05/21/resilient-ll-parsing-tutorial.html))
//! by matklad

// mod cst;
mod cst;
mod cursor;
mod grammar;
mod lexer;
mod parser;
mod token_kind;
mod tree_kind;
pub mod ungram;
mod version;

pub use cst::{ChildNodeKey, CstNode, CstNodeData, NodeId, SqliteToken, SqliteUntypedCst};

use enumset::EnumSet;
pub use token_kind::{sqlite_keywords, SqliteTokenKind, MAX_KEYWORD_LEN};

pub use tree_kind::SqliteTreeKind;

pub use version::SqliteVersion;

pub use parser::{SqliteParseError, SqliteParser};

pub use lexer::{LexError, SqliteLexer};

pub mod ast;

#[cfg(feature = "test_utils")]
pub mod test_utils;

pub fn parse(text: &str) -> SqliteUntypedCst {
    let (tokens, _) = SqliteLexer::new(text, SqliteVersion([3, 46, 0])).lex();

    let mut p = SqliteParser::new(tokens);
    grammar::file(&mut p, Default::default());

    let result = p.build_tree();

    result
}

pub fn parse_any(
    text: &str,
    r: EnumSet<SqliteTokenKind>,
    parse_function: fn(&mut SqliteParser, EnumSet<SqliteTokenKind>),
) -> SqliteUntypedCst {
    let (tokens, _) = SqliteLexer::new(text, SqliteVersion([3, 46, 0])).lex();
    let mut p = SqliteParser::new(tokens);
    parse_function(&mut p, r);

    p.build_tree()
}

#[test]
fn simple_parser_test() {
    let input = "WITH derived AS (
            SELECT MAX(a) AS max_a,
                   COUNT(b) AS b_num,
                   user_id
            FROM `TABLE`
            GROUP BY user_id
        )
        SELECT * FROM `table`
        LEFT JOIN derived USING (user_id);";

    let cst = parse(input);

    println!("{cst}");
    assert!(cst.errors().is_empty());
}
