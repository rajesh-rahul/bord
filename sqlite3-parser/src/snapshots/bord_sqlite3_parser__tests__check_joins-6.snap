---
source: sqlite3-parser/src/lib.rs
expression: cst
snapshot_kind: text
---
File(0..26)
└───JoinOperator(0..26)
    ├───Error: IllegalJoinOperator(0..21)
    ├   ├───KW_OUTER(0..5)
    ├   ├───WHITESPACE(5..6)
    ├   ├───KW_NATURAL(6..13)
    ├   ├───WHITESPACE(13..14)
    ├   └───KW_NATURAL(14..21)
    ├───WHITESPACE(21..22)
    └───KW_JOIN(22..26)
